### ISSUE TITLE
WebSocket: java.lang.IllegalStateException: already released RetainableByteBuffer

### ISSUE THREAD

**Issue by leonchen83** [eclipse/jetty.project#7086]
**Jetty version(s)**
10.0.7
**Java version/vendor** `(use: java -version)`
openjdk 11
**OS type/version**
Linux version 2.6.32-642.el6.x86_64 (mockbuild@worker1.bsys.centos.org) (gcc version 4.4.7 20120313 (Red Hat 4.4.7-17) (GCC) ) #1 SMP Tue May 10 17:27:01 UTC 2016
**Description**
occur exception on our prod env

```
AdaptiveExecutionStrategy.runTask:416]Task run failed
java.lang.IllegalStateException: already released RetainableByteBuffer@2c03deb4{DirectByteBuffer@6bbc693a[p=0,l=0,c=4096,r=0]={<<<>>>??ÖW?{i...},r=0}
at org.eclipse.jetty.io.RetainableByteBuffer.lambdarelease2(RetainableByteBuffer.java:109)
at java.base/java.util.concurrent.atomic.AtomicInteger.updateAndGet(AtomicInteger.java:279)
at org.eclipse.jetty.io.RetainableByteBuffer.release(RetainableByteBuffer.java:106)
at org.eclipse.jetty.websocket.core.internal.WebSocketConnection.releaseNetworkBuffer(WebSocketConnection.java:333)
at org.eclipse.jetty.websocket.core.internal.WebSocketConnection.fillAndParse(WebSocketConnection.java:507)
at org.eclipse.jetty.websocket.core.internal.WebSocketConnection.onFillable(WebSocketConnection.java:343)
at org.eclipse.jetty.io.AbstractConnectionReadCallback.succeeded(AbstractConnection.java:319)
at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
at org.eclipse.jetty.io.SocketChannelEndPoint1.run(SocketChannelEndPoint.java:101)
at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:412)
at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:381)
at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:268)
at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.lambdanew0(AdaptiveExecutionStrategy.java:138)
at org.eclipse.jetty.util.thread.ReservedThreadExecutorReservedThread.run(ReservedThreadExecutor.java:407)
at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:894)
at org.eclipse.jetty.util.thread.QueuedThreadPoolRunner.run(QueuedThreadPool.java:1038)
at java.base/java.lang.Thread.run(Thread.java:834)
```
_Do not report security issues here! See [Jetty Security Reports](https://www.eclipse.org/jetty/security_reports.php)._

**How to reproduce?**

it`s hard to reproduce

**Comment by lachlan-roberts** [comment]
There is not enough information to tell why this is happening.

@leonchen83 are you using the Jetty API or the Javax WebSocket API? and  what type of message handler are you using for this websocket connection?

**Comment by leonchen83** [comment]
we extend spring `WebSocketHandler` and handle text message with client
code like following
```
public class XWebSocketHandler implements WebSocketHandler {
....
....
    public final void handleMessage(final WebSocketSession session, final WebSocketMessage<?> message) throws Exception {
        if (message instanceof TextMessage) {
            // do our buinesses

        }
    }
}
```

**Comment by lachlan-roberts** [comment]
I think I've figured out why this is happening. 

The `WebSocketCoreSession.demand()` method will throw ISE if the session is no longer open. And this is typically done in a format like this:
```java
try
{
  ...
  callback.succeeded();
  coreSession.demand(1);
}
catch(Throwable t)
{
  callback.failed(t);
}
```

So if demand throws because the session is closed then it is both succeeding and failing the callback. You can trigger this by sending an error close frame from the `onMessage` method. See fix in PR #7088.
